lib/workspace.js

```diff
@@ -279,15 +280,58 @@ module.exports = {
     },
 
     async getGradedFilesFromS3(workspace_id) {
+        debug('Getting graded files from S3');
+        const result = await sqldb.queryOneRowAsync(sql.select_workspace_version_and_graded_files, { workspace_id });
+        const { workspace_version, workspace_graded_files } = result.rows[0];
         const timestamp = new Date().toISOString().replace(/[-T:.]/g, '-');
-        const zipName = `workspace-${workspace_id}-${timestamp}.zip`;
-        const zipPath = path.join(config.workspaceMainZipsDirectory, zipName);
-
-        // we should get the gradedFiles from S3 and zip them into zipPath
+        const zipName = `workspace-${workspace_id}-${workspace_version}-${timestamp}`;
+        const zipDir = path.join(config.workspaceMainZipsDirectory, zipName);
+        const zipPath = `${zipDir}.zip`;
 
-        throw new Error('getGradedFilesFromS3() is not yet implemented');
+        // download graded files from S3 -> zip dir -> zip file
+        // FIXME: stream straight from S3 -> zip file
+        const archive = archiver('zip');
+        const s3Name = `workspace-${workspace_id}-${workspace_version}`;
+        const zipJobs = workspace_graded_files.map(file => (
```

Comment:
You can just use https://caolan.github.io/async/v3/docs.html#mapLimit directly. Also the `async` lib can directly handle `async` functions (see https://caolan.github.io/async/v3/global.html) so there is no need to use `promisify()`.

Reply:
Combined into a single `eachLimit()`